plugin/healthchecker/filter.go

```diff
@@ -0,0 +1,127 @@
+package healthchecker
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru"
+	"github.com/miekg/dns"
+)
+
+type (
+	HealthCheckFilter struct {
+		cache    *lru.Cache
+		checker  Checker
+		interval time.Duration
+		names    map[string]struct{}
+	}
+
+	entry struct {
+		dnsRecord dns.RR
+		healthy   bool
+		quit      chan struct{}
+	}
+)
+
+var mtx sync.Mutex
+
+func NewHealthCheckFilter(checker Checker, size int, interval time.Duration, names map[string]struct{}) (*HealthCheckFilter, error) {
+	cache, err := lru.NewWithEvict(size, func(key interface{}, value interface{}) {
+		mtx.Lock()
+		if e, ok := value.(entry); ok {
+			close(e.quit)
+		}
+		mtx.Unlock()
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &HealthCheckFilter{
+		cache:    cache,
+		checker:  checker,
+		interval: interval,
+		names:    names,
+	}, nil
+}
+
+func (p *HealthCheckFilter) FilterRecords(records []dns.RR) []dns.RR {
+	if p.cache.Len() == 0 {
+		recs := make([]dns.RR, len(records))
+		copy(recs, records)
+		go func() {
+			for _, r := range recs {
+				if _, ok := p.names[r.Header().Name]; ok || len(p.names) == 0 {
+					p.put(r)
+				}
+			}
+		}()
+		return records
+	}
+
+	result := make([]dns.RR, 0, len(records))
+
+	for _, r := range records {
+		if _, ok := p.names[r.Header().Name]; ok || len(p.names) == 0 {
+			e := p.get(r.String())
+			if e != nil {
+				if e.healthy {
+					result = append(result, e.dnsRecord)
+				}
+				continue
+			}
+			p.put(r)
+		}
+		result = append(result, r)
+	}
+
+	return result
+}
+
+func (p *HealthCheckFilter) put(rec dns.RR) {
+	health := p.checker.Check(rec)
+	quit := make(chan struct{})
+	p.cache.Add(rec.String(), entry{
+		dnsRecord: rec,
+		healthy:   health,
+		quit:      quit,
+	})
+
+	ticker := time.NewTicker(p.interval)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-quit:
+				return
+			case <-ticker.C:
+				mtx.Lock()
+				if p.cache.Contains(rec.String()) {
+					p.cache.Add(rec.String(), entry{
```

Comment:
It seems we have problem with the mutex. During add new records eviction can happen (if cache is full) and we will be blocked. 
https://github.com/hashicorp/golang-lru/blob/master/simplelru/lru.go#L67
https://github.com/hashicorp/golang-lru/blob/80c98217689d6df152309d574ccc682b21dc802c/simplelru/lru.go#L162-L177

Sometimes could reproduce on the same previous test

/cc @alexvanin

Reply:
```
type tmpcheck struct {
}

func (t *tmpcheck) Check(record dns.RR) bool {
	return true
}

func TestPanic(t *testing.T) {
	checker := &tmpcheck{}

	f, err := NewHealthCheckFilter(checker, 2, 200 * time.Millisecond, map[string]struct{}{})
	require.NoError(t, err)

	a := &dns.A{A: net.ParseIP("127.0.0.1")}
	a2 := &dns.A{A: net.ParseIP("127.0.0.2")}
	a3 := &dns.A{A: net.ParseIP("127.0.0.3")}
	a4 := &dns.A{A: net.ParseIP("127.0.0.4")}

	f.put(a)
	f.put(a2)
	time.Sleep(200 * time.Millisecond)
	f.put(a3)
	time.Sleep(200 * time.Millisecond)
	f.put(a4)

	time.Sleep(time.Second)
}
```

Reply:
Using mutex per records instead of global mutex should solve this problem, I think.

Reply:
If cache will store pointers to `event` we can avoid usage of mutex, let's try this.